codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/MiddlewareStackSnapshotTests.java

```diff
@@ -0,0 +1,164 @@
+import software.amazon.smithy.go.codegen.SmithyGoDependency;
+import software.amazon.smithy.go.codegen.SmithyGoTypes;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.TopDownIndex;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.utils.MapUtils;
+
+public class MiddlewareStackSnapshotTests implements GoIntegration {
+    @Override
+    public void writeAdditionalFiles(
+            GoSettings settings, Model model, SymbolProvider symbolProvider, GoDelegator goDelegator
+    ) {
+        goDelegator.useFileWriter("snapshot_test.go", settings.getModuleName(), writer -> {
+            writer.addBuildTag("snapshot");
+            writer.write(commonTestSource());
+            writer.write(snapshotTests(model, settings.getService(model), symbolProvider));
+            writer.write(snapshotUpdaters(model, settings.getService(model), symbolProvider));
+        });
+    }
+
+    private GoWriter.Writable commonTestSource() {
+        return goTemplate("""
+                $os:D $fs:D $io:D $errors:D $fmt:D $middleware:D
+
+                const ssprefix = "snapshot"
+
+                type snapshotOK struct{}
+
+                func (snapshotOK) Error() string { return "error: success" }
+
+                func createp(path string) (*os.File, error) {
+                    if err := os.Mkdir(ssprefix, 0700); err != nil && !errors.Is(err, fs.ErrExist) {
+                        return nil, err
+                    }
+                    return os.Create(path)
+                }
+
+                func sspath(op string) string {
+                    return fmt.Sprintf("%s/api_op_%s.go.snap", ssprefix, op)
+                }
+
+                func updateSnapshot(stack *middleware.Stack, operation string) error {
+                    f, err := createp(sspath(operation))
+                    if err != nil {
+                        return err
+                    }
+                    defer f.Close()
+                    if _, err := f.Write([]byte(stack.String())); err != nil {
+                        return err
+                    }
+                    return snapshotOK{}
+                }
+
+                func testSnapshot(stack *middleware.Stack, operation string) error {
+                    f, err := os.Open(sspath(operation))
+                    if errors.Is(err, fs.ErrNotExist) {
+                        return snapshotOK{}
+                    }
+                    if err != nil {
+                        return err
+                    }
+                    defer f.Close()
+                    expected, err := io.ReadAll(f)
+                    if err != nil {
+                        return err
+                    }
+                    if actual := stack.String(); actual != string(expected) {
+                        return fmt.Errorf("%s != %s", expected, actual)
```

Comment:
I think it'd be better to use a dedicated diff tool, but that can be an incremental improvement.